src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/CollectionExpressionExtensions.cs

```diff
@@ -0,0 +1,178 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Shared.Extensions;
+
+internal static class CollectionExpressionExtensions
+{
+    public static bool IsWellKnownCollectionInterface(ITypeSymbol type)
+        => IsWellKnownCollectionReadOnlyInterface(type) || IsWellKnownCollectionReadWriteInterface(type);
+
+    public static bool IsWellKnownCollectionReadOnlyInterface(ITypeSymbol type)
+    {
+        return type.OriginalDefinition.SpecialType
+            is SpecialType.System_Collections_Generic_IEnumerable_T
+            or SpecialType.System_Collections_Generic_IReadOnlyCollection_T
+            or SpecialType.System_Collections_Generic_IReadOnlyList_T;
+    }
+
+    public static bool IsWellKnownCollectionReadWriteInterface(ITypeSymbol type)
+    {
+        return type.OriginalDefinition.SpecialType
+            is SpecialType.System_Collections_Generic_ICollection_T
+            or SpecialType.System_Collections_Generic_IList_T;
+    }
+
+    public static bool IsConstructibleCollectionType(
+        Compilation compilation,
+        [NotNullWhen(true)] ITypeSymbol? type)
+    {
+        return IsConstructibleCollectionType(compilation, type, out _);
+    }
+
+    public static bool IsConstructibleCollectionType(
+        Compilation compilation,
+        [NotNullWhen(true)] ITypeSymbol? type,
+        out ImmutableArray<ITypeSymbol> elementTypes)
+    {
+        if (type is null)
+        {
+            elementTypes = default;
+            return false;
+        }
+
+        // Arrays are always a valid collection expression type.
+        if (type is IArrayTypeSymbol arrayType)
+        {
+            elementTypes = [arrayType.ElementType];
+            return true;
+        }
+
+        // Has to be a real named type at this point.
+        if (type is INamedTypeSymbol namedType)
+        {
+            // Span<T> and ReadOnlySpan<T> are always valid collection expression types.
+            if (namedType.OriginalDefinition.Equals(compilation.SpanOfTType()) ||
+                namedType.OriginalDefinition.Equals(compilation.ReadOnlySpanOfTType()))
+            {
+                elementTypes = namedType.TypeArguments;
+                return true;
+            }
+
+            // If it has a [CollectionBuilder] attribute on it, it is a valid collection expression type.
+            var collectionBuilderMethods = CollectionExpressionExtensions.TryGetCollectionBuilderFactoryMethods(
```

Comment:
```suggestion
            var collectionBuilderMethods = TryGetCollectionBuilderFactoryMethods(
```

Not sure why qualified here.

Reply:
removed.